service/header/store.go

```diff
@@ -0,0 +1,302 @@
+	if v, ok := s.cache.Get(hash); ok {
+		return v.(*ExtendedHeader), nil
+	}
+
+	b, err := s.ds.Get(datastore.NewKey(hash.String()))
+	if err != nil {
+		if err == datastore.ErrNotFound {
+			return nil, ErrNotFound
+		}
+
+		return nil, err
+	}
+
+	return UnmarshalExtendedHeader(b)
+}
+
+func (s *store) GetByHeight(ctx context.Context, height int64) (*ExtendedHeader, error) {
+	hash, err := s.index.HashByHeight(height)
+	if err != nil {
+		if err == datastore.ErrNotFound {
+			return nil, ErrNotFound
+		}
+
+		return nil, err
+	}
+
+	return s.Get(ctx, hash)
+}
+
+func (s *store) GetRangeByHeight(ctx context.Context, from, to int64) ([]*ExtendedHeader, error) {
+	h, err := s.GetByHeight(ctx, to)
+	if err != nil {
+		return nil, err
+	}
+
+	ln := to - from
+	headers := make([]*ExtendedHeader, ln)
+	for i := ln - 1; i > 0; i-- {
+		headers[i] = h
+		h, err = s.Get(ctx, h.LastHeader())
+		if err != nil {
+			return nil, err
+		}
+	}
+	headers[0] = h
+
+	return headers, nil
+}
+
+func (s *store) Has(_ context.Context, hash bytes.HexBytes) (bool, error) {
+	if ok := s.cache.Contains(hash); ok {
+		return ok, nil
+	}
+
+	return s.ds.Has(datastore.NewKey(hash.String()))
+}
+
+func (s *store) Append(ctx context.Context, headers ...*ExtendedHeader) error {
+	lh := len(headers)
+	if lh == 0 {
+		return nil
+	}
+
+	head, err := s.Head(ctx)
+	if err != nil {
+		return err
+	}
+
+	verified := make([]*ExtendedHeader, 0, lh)
+	for _, h := range headers {
+		err = VerifyAdjacent(head, h)
+		if err != nil {
+			log.Errorw("invalid header", "head", head.Hash(), "header", h.Hash(), "err", err)
+			break // if some headers are cryptographically valid, why not include them? Exactly, so let's include
+		}
+		verified, head = append(verified, h), h
+	}
+	if len(verified) == 0 {
+		return fmt.Errorf("header/store: no valid headers were given")
+	}
+
+	err = s.put(verified...)
+	if err != nil {
+		return err
+	}
+
+	// consistency is important, so change the cache and the head only after the data is on disk
+	for _, h := range verified {
+		s.cache.Add(h.Hash(), h)
+	}
+
+	return s.newHead(head.Hash())
+}
+
+func (s *store) put(headers ...*ExtendedHeader) error {
+	batch, err := s.ds.Batch()
+	if err != nil {
+		return err
+	}
+
+	for _, h := range headers {
+		b, err := h.MarshalBinary()
+		if err != nil {
+			return err
+		}
+
+		err = batch.Put(headerKey(h), b)
+		if err != nil {
+			return err
+		}
+	}
+
+	err = batch.Commit()
+	if err != nil {
+		return err
+	}
+
+	return s.index.Index(headers...)
+}
+
+func (s *store) loadHead() error {
+	b, err := s.ds.Get(headKey)
+	if err != nil {
+		return err
+	}
+
+	s.headLk.Lock()
+	defer s.headLk.Unlock()
+	err = s.head.UnmarshalJSON(b)
+	if err != nil {
+		return err
+	}
+
+	log.Infow("loaded head", "hash", s.head)
+	return nil
+}
+
+func (s *store) newHead(head bytes.HexBytes) error {
+	log.Infow("new head", "hash", head)
+
+	// at this point Header body of the given 'head' must be already written with put
```

Comment:
you can mention this in the docs of this method

Reply:
Ok